File: android-vinetrack/app/src/main/java/com/rork/vinetrack/data/WorkTaskLabourCosting.kt
package com.rork.vinetrack.data

import com.rork.vinetrack.data.model.OperatorCategory
import com.rork.vinetrack.data.model.WorkTaskLabourLine

/**
 * THE authoritative labour-costing contract for Work Tasks — the Kotlin twin of
 * the Swift `WorkTaskLabourCosting`. Both platforms MUST produce identical
 * numbers from identical inputs; the unit suites on each side assert the same
 * fixtures.
 *
 * Ownership rule this object encodes:
 *
 *  * A **Work Task labour line** owns labour type, hourly rate, worker count,
 *    hours per worker, person-hours and labour cost. These are the ONLY
 *    authoritative source of labour cost.
 *  * A **Pruning Activity** owns operational output (date, blocks, rows,
 *    quarters, method, crew reference, vines, start/finish, elapsed duration,
 *    notes, `work_task_id`). It no longer offers an editable hourly rate.
 *
 * Legacy `pruning_activities.hourly_rate` / `labour_hours` values stay READABLE
 * for compatibility, but they are never combined with Work Task labour totals —
 * see [resolveLabour], which returns exactly one source.
 *
 * Arithmetic (mirrored in Swift):
 *
 *     person-hours = workerCount × hoursPerWorker
 *     line cost    = person-hours × hourlyRate
 *     task totals  = Σ line person-hours, Σ line costs
 *
 * An already-aggregated activity duration is NEVER multiplied again: elapsed
 * activity duration, hours per worker, and total person-hours are three
 * distinct quantities.
 */
object WorkTaskLabourCosting {

    // MARK: Bounds (shared with the standard Work Task form)

    /** A crew larger than this is a typo, not a vineyard. */
    const val MAX_WORKER_COUNT: Int = 500

    /** More than a full day per worker on one line is a typo. */
    const val MAX_HOURS_PER_WORKER: Double = 24.0

    /** Defensive upper bound on an hourly rate. */
    const val MAX_HOURLY_RATE: Double = 100_000.0

    // MARK: Line arithmetic

    /** `workerCount × hoursPerWorker`, floored at zero and never NaN/∞. */
    fun personHours(workerCount: Int, hoursPerWorker: Double): Double {
        if (!hoursPerWorker.isFinite()) return 0.0
        val count = workerCount.coerceAtLeast(0)
        val hours = hoursPerWorker.coerceAtLeast(0.0)
        return count.toDouble() * hours
    }

    /** `person-hours × hourlyRate`. A null rate means "cost not specified". */
    fun lineCost(workerCount: Int, hoursPerWorker: Double, hourlyRate: Double?): Double? {
        val rate = hourlyRate ?: return null
        if (!rate.isFinite()) return null
        return personHours(workerCount, hoursPerWorker) * rate.coerceAtLeast(0.0)
    }

    /** Person-hours of one stored line, preferring the DB-generated value. */
    fun personHours(line: WorkTaskLabourLine): Double {
        val stored = line.totalHours
        if (stored != null && stored.isFinite() && stored >= 0) return stored
        return personHours(line.workerCount, line.hoursPerWorker)
    }

    /**
     * Cost of one stored line. Null when neither a stored total nor a rate
     * exists — an absent rate is "not specified", never `$0.00`.
     */
    fun lineCost(line: WorkTaskLabourLine): Double? {
        val stored = line.totalCost
        if (stored != null && stored.isFinite()) return stored
        val rate = line.hourlyRate ?: return null
        if (!rate.isFinite()) return null
        return personHours(line) * rate.coerceAtLeast(0.0)
    }

    // MARK: Task totals

    /** Live (non-deleted) lines of one task, oldest first. */
    fun linesFor(lines: List<WorkTaskLabourLine>, workTaskId: String): List<WorkTaskLabourLine> =
        lines
            .filter { it.workTaskId == workTaskId && it.deletedAt == null }
            .sortedBy { it.workDate.orEmpty() }

    /** Σ person-hours across the given lines. */
    fun totalPersonHours(lines: List<WorkTaskLabourLine>): Double =
        lines.sumOf { personHours(it) }

    /**
     * Σ costs across the given lines. Null when NO line carries a cost at all,
     * so "not specified" never renders as zero. Lines without a rate contribute
     * nothing rather than blocking the total.
     */
    fun totalCost(lines: List<WorkTaskLabourLine>): Double? {
        var sum = 0.0
        var sawCost = false
        for (line in lines) {
            val cost = lineCost(line) ?: continue
            sum += cost
            sawCost = true
        }
        return if (sawCost) sum else null
    }

    /** Σ worker counts across the given lines. */
    fun totalWorkers(lines: List<WorkTaskLabourLine>): Int = lines.sumOf { it.workerCount.coerceAtLeast(0) }

    /** One task's totals in a single pass. */
    fun totals(lines: List<WorkTaskLabourLine>): LabourTotals = LabourTotals(
        lineCount = lines.size,
        workers = totalWorkers(lines),
        personHours = totalPersonHours(lines),
        cost = totalCost(lines),
    )

    // MARK: Legacy fallback (never double-counted)

    /**
     * Which source a pruning report must use for one activity's labour.
     *
     *  * [PIECE_RATE] — the linked task is costed per vine (sql/188). Its own
     *    snapshot IS the labour-cost record, so it wins outright and labour
     *    lines on the same task are operational history only.
     *  * [WORK_TASK_LINES] — the linked task has labour lines. Authoritative.
     *  * [LEGACY_ACTIVITY] — no labour lines exist; the activity's own historical
     *    `labour_hours` / `hourly_rate` are shown as legacy data.
     *  * [NONE] — nothing recorded.
     *
     * The value sources are mutually exclusive by construction, which is what
     * stops a report summing an activity rate AND its task's lines — or a
     * piece-rate total AND an hourly one.
     */
    enum class LabourSource { PIECE_RATE, WORK_TASK_LINES, LEGACY_ACTIVITY, NONE }

    data class LabourTotals(
        val lineCount: Int,
        val workers: Int,
        val personHours: Double,
        val cost: Double?,
    ) {
        val isEmpty: Boolean get() = lineCount == 0
    }

    data class ResolvedLabour(
        val source: LabourSource,
        /** Person-hours from labour lines, or the legacy activity hours. */
        val hours: Double?,
        /** Cost from labour lines, or the legacy activity hours × rate. */
        val cost: Double?,
        /** True when [hours]/[cost] come from pre-labour-line history. */
        val isLegacy: Boolean,
    )

    /**
     * Resolves ONE activity's labour, preferring Work Task labour lines and
     * falling back to the legacy activity-level values only when the task has no
     * lines at all. Exactly one source contributes, so no total can count labour
     * twice.
     *
     * @param lines every live labour line of the LINKED task (empty when the
     *   activity has no task, or the task has no lines).
     * @param legacyHours historical `pruning_activities.labour_hours`.
     * @param legacyRate historical `pruning_activities.hourly_rate`.
     * @param includeCost false hides every monetary value for roles without
     *   costing visibility (hours are still returned).
     */
    fun resolveLabour(
        lines: List<WorkTaskLabourLine>,
        legacyHours: Double?,
        legacyRate: Double?,
        includeCost: Boolean = true,
    ): ResolvedLabour {
        val live = lines.filter { it.deletedAt == null }
        if (live.isNotEmpty()) {
            val hours = totalPersonHours(live)
            return ResolvedLabour(
                source = LabourSource.WORK_TASK_LINES,
                hours = hours.takeIf { it > 0 },
                cost = if (includeCost) totalCost(live) else null,
                isLegacy = false,
            )
        }
        val hours = legacyHours?.takeIf { it.isFinite() && it > 0 }
        val rate = legacyRate?.takeIf { it.isFinite() && it > 0 }
        if (hours == null && rate == null) {
            return ResolvedLabour(LabourSource.NONE, null, null, isLegacy = false)
        }
        return ResolvedLabour(
            source = LabourSource.LEGACY_ACTIVITY,
            hours = hours,
            cost = if (includeCost && hours != null && rate != null) hours * rate else null,
            isLegacy = true,
        )
    }

    /**
     * Per-task labour cost map from labour lines ONLY.
     *
     * This answers "what do this task's labour lines cost?", NOT "what is this
     * task's labour cost?". A piece-rate task legitimately has no labour lines,
     * so any caller that means the latter must use
     * [com.rork.vinetrack.data.model.PieceRateCosting.effectiveCostsByWorkTask],
     * which applies the `costing_method` switch first.
     */
    fun costsByWorkTask(
        lines: List<WorkTaskLabourLine>,
        includeCost: Boolean = true,
    ): Map<String, Double> {
        if (!includeCost) return emptyMap()
        return lines
            .filter { it.deletedAt == null }
            .groupBy { it.workTaskId }
            .mapNotNull { (taskId, group) -> totalCost(group)?.let { taskId to it } }
            .toMap()
    }

    /** Per-task person-hours map for the Activity Report, built in ONE pass. */
    fun hoursByWorkTask(lines: List<WorkTaskLabourLine>): Map<String, Double> = lines
        .filter { it.deletedAt == null }
        .groupBy { it.workTaskId }
        .mapValues { (_, group) -> totalPersonHours(group) }
        .filterValues { it > 0 }

    // MARK: Defaults

    /**
     * The rate a newly selected labour type supplies. The saved worker-type
     * cost is the default; an explicit edit always wins.
     */
    fun defaultRate(category: OperatorCategory?): Double? =
        category?.costPerHour?.takeIf { it.isFinite() && it > 0 }

    // MARK: Validation

    enum class LabourLineField { LABOUR_TYPE, WORKER_COUNT, HOURS_PER_WORKER, HOURLY_RATE }

    data class LabourLineIssue(val field: LabourLineField, val message: String)

    /**
     * Inline validation for one labour line. Returns EVERY problem so the form
     * can mark each field; an empty list means saveable. The caller must keep
     * the entered values on failure — nothing here discards input.
     *
     * Rules (identical on iOS):
     *  * a labour type is required (a linked worker type or a typed name),
     *  * worker count > 0 and ≤ [MAX_WORKER_COUNT],
     *  * hours per worker > 0 and ≤ [MAX_HOURS_PER_WORKER],
     *  * hourly rate, when given, ≥ 0 and ≤ [MAX_HOURLY_RATE],
     *  * every numeric value must be finite.
     */
    fun validate(
        labourType: String,
        workerCount: Int?,
        hoursPerWorker: Double?,
        hourlyRate: Double?,
        rateProvided: Boolean = hourlyRate != null,
    ): List<LabourLineIssue> {
        val issues = mutableListOf<LabourLineIssue>()

        if (labourType.trim().isEmpty()) {
            issues += LabourLineIssue(LabourLineField.LABOUR_TYPE, "Choose a labour type.")
        }

        when {
            workerCount == null ->
                issues += LabourLineIssue(LabourLineField.WORKER_COUNT, "Enter how many people worked.")
            workerCount <= 0 ->
                issues += LabourLineIssue(LabourLineField.WORKER_COUNT, "At least one person is needed.")
            workerCount > MAX_WORKER_COUNT ->
                issues += LabourLineIssue(
                    LabourLineField.WORKER_COUNT,
                    "That is more than $MAX_WORKER_COUNT people — check the number.",
                )
        }

        when {
            hoursPerWorker == null || !hoursPerWorker.isFinite() ->
                issues += LabourLineIssue(LabourLineField.HOURS_PER_WORKER, "Enter the hours each person worked.")
            hoursPerWorker <= 0.0 ->
                issues += LabourLineIssue(LabourLineField.HOURS_PER_WORKER, "Hours must be more than zero.")
            hoursPerWorker > MAX_HOURS_PER_WORKER ->
                issues += LabourLineIssue(
                    LabourLineField.HOURS_PER_WORKER,
                    "One person cannot work more than ${MAX_HOURS_PER_WORKER.toInt()} hours in a day.",
                )
        }

        if (rateProvided) {
            when {
                hourlyRate == null || !hourlyRate.isFinite() ->
                    issues += LabourLineIssue(LabourLineField.HOURLY_RATE, "Enter a valid hourly rate.")
                hourlyRate < 0.0 ->
                    issues += LabourLineIssue(LabourLineField.HOURLY_RATE, "An hourly rate cannot be negative.")
                hourlyRate > MAX_HOURLY_RATE ->
                    issues += LabourLineIssue(LabourLineField.HOURLY_RATE, "That hourly rate looks too high.")
            }
        }

        return issues
    }

    fun isValid(
        labourType: String,
        workerCount: Int?,
        hoursPerWorker: Double?,
        hourlyRate: Double?,
        rateProvided: Boolean = hourlyRate != null,
    ): Boolean = validate(labourType, workerCount, hoursPerWorker, hourlyRate, rateProvided).isEmpty()

    fun message(issues: List<LabourLineIssue>, field: LabourLineField): String? =
        issues.firstOrNull { it.field == field }?.message
}

File: ios/VineTrack/LegacyImported/Models/WorkTaskLabourCosting.swift
import Foundation

/// THE authoritative labour-costing contract for Work Tasks — the Swift twin of
/// the Kotlin `WorkTaskLabourCosting`. Both platforms MUST produce identical
/// numbers from identical inputs; the unit suites on each side assert the same
/// fixtures.
///
/// Ownership rule this type encodes:
///
/// * A **Work Task labour line** owns labour type, hourly rate, worker count,
///   hours per worker, person-hours and labour cost. These are the ONLY
///   authoritative source of labour cost.
/// * A **Pruning Activity** owns operational output (date, blocks, rows,
///   quarters, method, crew reference, vines, start/finish, elapsed duration,
///   notes, `work_task_id`). It no longer offers an editable hourly rate.
///
/// Legacy `pruning_activities.hourly_rate` / `labour_hours` values stay READABLE
/// for compatibility, but they are never combined with Work Task labour totals —
/// see `resolveLabour(...)`, which returns exactly one source.
///
/// Arithmetic (mirrored in Kotlin):
///
///     person-hours = workerCount × hoursPerWorker
///     line cost    = person-hours × hourlyRate
///     task totals  = Σ line person-hours, Σ line costs
///
/// An already-aggregated activity duration is NEVER multiplied again: elapsed
/// activity duration, hours per worker, and total person-hours are three
/// distinct quantities.
nonisolated enum WorkTaskLabourCosting {

    // MARK: Bounds (shared with the standard Work Task form)

    /// A crew larger than this is a typo, not a vineyard.
    static let maxWorkerCount: Int = 500

    /// More than a full day per worker on one line is a typo.
    static let maxHoursPerWorker: Double = 24

    /// Defensive upper bound on an hourly rate.
    static let maxHourlyRate: Double = 100_000

    // MARK: Line arithmetic

    /// `workerCount × hoursPerWorker`, floored at zero and never NaN/∞.
    static func personHours(workerCount: Int, hoursPerWorker: Double) -> Double {
        guard hoursPerWorker.isFinite else { return 0 }
        return Double(max(workerCount, 0)) * max(hoursPerWorker, 0)
    }

    /// `person-hours × hourlyRate`. A nil rate means "cost not specified".
    static func lineCost(workerCount: Int, hoursPerWorker: Double, hourlyRate: Double?) -> Double? {
        guard let hourlyRate, hourlyRate.isFinite else { return nil }
        return personHours(workerCount: workerCount, hoursPerWorker: hoursPerWorker) * max(hourlyRate, 0)
    }

    /// Person-hours of one stored line, preferring the DB-generated value.
    static func personHours(_ line: WorkTaskLabourLine) -> Double {
        personHours(workerCount: line.workerCount, hoursPerWorker: line.hoursPerWorker)
    }

    /// Cost of one stored line. Nil when no rate exists — an absent rate is
    /// "not specified", never `$0.00`.
    static func lineCost(_ line: WorkTaskLabourLine) -> Double? {
        guard let rate = line.hourlyRate, rate.isFinite else { return nil }
        return personHours(line) * max(rate, 0)
    }

    // MARK: Task totals

    /// Live lines of one task, oldest first.
    static func lines(_ lines: [WorkTaskLabourLine], for workTaskId: UUID) -> [WorkTaskLabourLine] {
        lines
            .filter { $0.workTaskId == workTaskId }
            .sorted { $0.workDate < $1.workDate }
    }

    /// Σ person-hours across the given lines.
    static func totalPersonHours(_ lines: [WorkTaskLabourLine]) -> Double {
        lines.reduce(0) { $0 + personHours($1) }
    }

    /// Σ costs across the given lines. Nil when NO line carries a cost at all,
    /// so "not specified" never renders as zero. Lines without a rate contribute
    /// nothing rather than blocking the total.
    static func totalCost(_ lines: [WorkTaskLabourLine]) -> Double? {
        var sum: Double = 0
        var sawCost = false
        for line in lines {
            guard let cost = lineCost(line) else { continue }
            sum += cost
            sawCost = true
        }
        return sawCost ? sum : nil
    }

    /// Σ worker counts across the given lines.
    static func totalWorkers(_ lines: [WorkTaskLabourLine]) -> Int {
        lines.reduce(0) { $0 + max($1.workerCount, 0) }
    }

    nonisolated struct LabourTotals: Equatable, Sendable {
        var lineCount: Int
        var workers: Int
        var personHours: Double
        var cost: Double?

        var isEmpty: Bool { lineCount == 0 }
    }

    /// One task's totals in a single pass.
    static func totals(_ lines: [WorkTaskLabourLine]) -> LabourTotals {
        LabourTotals(
            lineCount: lines.count,
            workers: totalWorkers(lines),
            personHours: totalPersonHours(lines),
            cost: totalCost(lines)
        )
    }

    // MARK: Legacy fallback (never double-counted)

    /// Which source a pruning report must use for one activity's labour.
    ///
    /// * `pieceRate` — the linked task is costed per vine (sql/188). Its own
    ///   snapshot IS the labour-cost record, so it wins outright and labour
    ///   lines on the same task are operational history only.
    /// * `workTaskLines` — the linked task has labour lines. Authoritative.
    /// * `legacyActivity` — no labour lines exist; the activity's own historical
    ///   `labour_hours` / `hourly_rate` are shown as legacy data.
    /// * `none` — nothing recorded.
    ///
    /// The value sources are mutually exclusive by construction, which is what
    /// stops a report summing an activity rate AND its task's lines — or a
    /// piece-rate total AND an hourly one.
    nonisolated enum LabourSource: String, Sendable {
        case pieceRate
        case workTaskLines
        case legacyActivity
        case none
    }

    nonisolated struct ResolvedLabour: Equatable, Sendable {
        var source: LabourSource
        /// Person-hours from labour lines, or the legacy activity hours.
        var hours: Double?
        /// Cost from labour lines, or the legacy activity hours × rate.
        var cost: Double?
        /// True when `hours`/`cost` come from pre-labour-line history.
        var isLegacy: Bool
    }

    /// Resolves ONE activity's labour, preferring Work Task labour lines and
    /// falling back to the legacy activity-level values only when the task has
    /// no lines at all. Exactly one source contributes, so no total can count
    /// labour twice.
    ///
    /// - Parameters:
    ///   - lines: every live labour line of the LINKED task (empty when the
    ///     activity has no task, or the task has no lines).
    ///   - legacyHours: historical `pruning_activities.labour_hours`.
    ///   - legacyRate: historical `pruning_activities.hourly_rate`.
    ///   - includeCost: false hides every monetary value for roles without
    ///     costing visibility (hours are still returned).
    static func resolveLabour(
        lines: [WorkTaskLabourLine],
        legacyHours: Double?,
        legacyRate: Double?,
        includeCost: Bool = true
    ) -> ResolvedLabour {
        if !lines.isEmpty {
            let hours = totalPersonHours(lines)
            return ResolvedLabour(
                source: .workTaskLines,
                hours: hours > 0 ? hours : nil,
                cost: includeCost ? totalCost(lines) : nil,
                isLegacy: false
            )
        }
        let hours: Double? = (legacyHours?.isFinite == true && (legacyHours ?? 0) > 0) ? legacyHours : nil
        let rate: Double? = (legacyRate?.isFinite == true && (legacyRate ?? 0) > 0) ? legacyRate : nil
        if hours == nil && rate == nil {
            return ResolvedLabour(source: .none, hours: nil, cost: nil, isLegacy: false)
        }
        var cost: Double?
        if includeCost, let hours, let rate { cost = hours * rate }
        return ResolvedLabour(source: .legacyActivity, hours: hours, cost: cost, isLegacy: true)
    }

    /// Per-task labour cost map from labour lines ONLY.
    ///
    /// - Warning: this answers "what do this task's labour lines cost?", NOT
    ///   "what is this task's labour cost?". A piece-rate task legitimately has
    ///   no labour lines, so any caller that means the latter must use
    ///   `PieceRateCosting.effectiveCostsByWorkTask(tasks:labourLines:includeCost:)`,
    ///   which applies the `costing_method` switch first.
    static func costsByWorkTask(_ lines: [WorkTaskLabourLine], includeCost: Bool = true) -> [UUID: Double] {
        guard includeCost else { return [:] }
        var grouped: [UUID: [WorkTaskLabourLine]] = [:]
        for line in lines { grouped[line.workTaskId, default: []].append(line) }
        var costs: [UUID: Double] = [:]
        for (taskId, group) in grouped {
            if let cost = totalCost(group) { costs[taskId] = cost }
        }
        return costs
    }

    /// Per-task person-hours map for the Activity Report, built in ONE pass.
    static func hoursByWorkTask(_ lines: [WorkTaskLabourLine]) -> [UUID: Double] {
        var hours: [UUID: Double] = [:]
        for line in lines { hours[line.workTaskId, default: 0] += personHours(line) }
        return hours.filter { $0.value > 0 }
    }

    // MARK: Defaults

    /// The rate a newly selected labour type supplies. The saved worker-type
    /// cost is the default; an explicit edit always wins.
    static func defaultRate(_ category: OperatorCategory?) -> Double? {
        guard let cost = category?.costPerHour, cost.isFinite, cost > 0 else { return nil }
        return cost
    }

    // MARK: Validation

    nonisolated enum LabourLineField: String, Sendable, Hashable {
        case labourType
        case workerCount
        case hoursPerWorker
        case hourlyRate
    }

    nonisolated struct LabourLineIssue: Equatable, Sendable {
        var field: LabourLineField
        var message: String
    }

    /// Inline validation for one labour line. Returns EVERY problem so the form
    /// can mark each field; an empty array means saveable. The caller must keep
    /// the entered values on failure — nothing here discards input.
    ///
    /// Rules (identical on Android):
    /// * a labour type is required (a linked worker type or a typed name),
    /// * worker count > 0 and ≤ `maxWorkerCount`,
    /// * hours per worker > 0 and ≤ `maxHoursPerWorker`,
    /// * hourly rate, when given, ≥ 0 and ≤ `maxHourlyRate`,
    /// * every numeric value must be finite.
    static func validate(
        labourType: String,
        workerCount: Int?,
        hoursPerWorker: Double?,
        hourlyRate: Double?,
        rateProvided: Bool? = nil
    ) -> [LabourLineIssue] {
        var issues: [LabourLineIssue] = []

        if labourType.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
            issues.append(LabourLineIssue(field: .labourType, message: "Choose a labour type."))
        }

        if let workerCount {
            if workerCount <= 0 {
                issues.append(LabourLineIssue(field: .workerCount, message: "At least one person is needed."))
            } else if workerCount > maxWorkerCount {
                issues.append(LabourLineIssue(
                    field: .workerCount,
                    message: "That is more than \(maxWorkerCount) people — check the number."
                ))
            }
        } else {
            issues.append(LabourLineIssue(field: .workerCount, message: "Enter how many people worked."))
        }

        if let hoursPerWorker, hoursPerWorker.isFinite {
            if hoursPerWorker <= 0 {
                issues.append(LabourLineIssue(field: .hoursPerWorker, message: "Hours must be more than zero."))
            } else if hoursPerWorker > maxHoursPerWorker {
                issues.append(LabourLineIssue(
                    field: .hoursPerWorker,
                    message: "One person cannot work more than \(Int(maxHoursPerWorker)) hours in a day."
                ))
            }
        } else {
            issues.append(LabourLineIssue(
                field: .hoursPerWorker,
                message: "Enter the hours each person worked."
            ))
        }

        if rateProvided ?? (hourlyRate != nil) {
            if let hourlyRate, hourlyRate.isFinite {
                if hourlyRate < 0 {
                    issues.append(LabourLineIssue(field: .hourlyRate, message: "An hourly rate cannot be negative."))
                } else if hourlyRate > maxHourlyRate {
                    issues.append(LabourLineIssue(field: .hourlyRate, message: "That hourly rate looks too high."))
                }
            } else {
                issues.append(LabourLineIssue(field: .hourlyRate, message: "Enter a valid hourly rate."))
            }
        }

        return issues
    }

    static func isValid(
        labourType: String,
        workerCount: Int?,
        hoursPerWorker: Double?,
        hourlyRate: Double?,
        rateProvided: Bool? = nil
    ) -> Bool {
        validate(
            labourType: labourType,
            workerCount: workerCount,
            hoursPerWorker: hoursPerWorker,
            hourlyRate: hourlyRate,
            rateProvided: rateProvided
        ).isEmpty
    }

    static func message(_ issues: [LabourLineIssue], for field: LabourLineField) -> String? {
        issues.first { $0.field == field }?.message
    }
}
